sycl/include/sycl/ext/intel/esimd/detail/atomic_intrin.hpp

```diff
@@ -39,7 +45,21 @@ template <typename Ty> Ty atomic_add_fetch(Ty *ptr, Ty val) {
   // TODO: Windows will be supported soon
   __ESIMD_UNSUPPORTED_ON_HOST;
 #else
-  return __atomic_add_fetch(ptr, val, __ATOMIC_SEQ_CST);
+  if constexpr (std::is_integral_v<Ty>) {
+    return __atomic_add_fetch(ptr, val, __ATOMIC_SEQ_CST);
+  } else {
+    // For Floating type
+    Ty _old, _new;
+    CmpxchgTy<Ty> _old_bits, _new_bits;
+    do {
+      _old = *ptr;
+      _new = _old + val;
+      _old_bits = *(CmpxchgTy<Ty> *)&_old;
+      _new_bits = *(CmpxchgTy<Ty> *)&_new;
```

Comment:
Nit: these two can be moved out of the loop (maybe in a separate patch)

Reply:
Will do later in another PR.